### PR TITLE
Extract reusable GPU Sim and renderer Swift packages

### PR DESCRIPTION
**Pull request by LuckyIYI**
## Summary

- extract the simulator into standalone, multi-backend-ready SwiftPM products behind the `GPUSim` facade
- add the optional `GPUSimRenderer` product with a backend-neutral scene protocol and a compact `MTKView` integration API
- keep application capture and debug policy outside the reusable renderer module
- migrate the development apps and add external consumer fixtures, documentation, tests, and architecture boundary checks

## Validation

- root package tests passed: 265 XCTest tests plus 2 Swift Testing tests; the final renderer delta passed all 4 Metal tests
- Development package tests passed: 427 tests with 20 expected optional-backend skips
- root and Development release builds passed
- standalone simulation and renderer consumer fixtures passed
- generic arm64 iOS package build passed
- architecture verifier and its 22 tests passed
- visual renderer smoke test passed on real Metal hardware

### CODE REVIEW

**Comment by LuckyIYI** [comment]
Reviewed and tested by porting a real consumer: the FrankaTeleop game (arm + puzzle generator + phone teleop) now builds and runs as an **external SwiftPM consumer** of this branch, alongside teleopkit. Verification first, then the API feedback that fell out of the port.

## Verification

- Root package: `swift build` + `swift test` clean on Apple silicon (exit 0, including the Metal suites).
- Ported `PuzzleGen` + `FrankaTeleop` to a standalone package depending on `GPUSim`/`GPUSimRenderer` by path. The **entire migration for the simulator side was an import swap** (`import SimCore` + `import PhysicsAVBD` → `import GPUSim`) — the `@_exported` facade works exactly as advertised, ~30 files, zero code changes.
- Game gates on this branch's engine, through public products only: `--selftest` PASSED (full scripted pick-and-place, solver 3.8 ms/tick), `--remotetest` PASSED (33 teleop/control checks over a live UDP loopback).
- `GPUSimRenderer` driving the live game world (`GPUSimRendererSource` wrapping the fixed-timestep world tick + phone pump): renders the Panda, gripper, puzzle, shadows, GTAO at 60 Hz with an iPhone teleoperating in the loop. `frameCompletionHandler` made a headless frame-capture check trivial — no screen-recording permission dance. Nice touch.

## Missing public API (blocking for this consumer)

`PuzzleGen`'s gate predicates and the game's round reset need per-joint fracture access; the branch only exposes the `debugBrokenJoints()` count. We had to add these two (lifted from the teleop branch's engine) to complete the port:

- `public func brokenJointIndices() -> [Int]` — which breakable joints have fractured (`header.z`).
- `public func repairJoints(_ jointIndices: [Int]? = nil)` — re-arm fractured joints, clearing warm starts/penalties like a pose reset does. Fracture is one-way in the solver, so any scene-reset flow (our `newRound`) silently leaves bonds broken without it — we hit exactly this as a live bug last week.

Suggest including both in the `GPUSim` public surface (and the API-compatibility tests).

## Renderer feature requests (what separates this from replacing our custom renderer)

The game still ships a 1300-line bespoke Metal renderer only because of these; with them we'd delete it:

1. **App-owned camera pose.** The orbit parameters cover editor-style viewing, but teleop needs egocentric cameras: a wrist camera rigidly mounted on the palm and an over-the-hand chase view. A `setCamera(viewMatrix:)`/`(position:target:up:)` override (with `automaticallyFramesScene = false`) would cover it — everything else (projection, temporal reprojection reset) already exists.
2. **A second-pass / overlay extension point.** Apps draw non-physics chrome into the scene: alignment guides, drop-target discs, ghost poses, translucent previews. Either (a) an app-supplied auxiliary `GPUSimRenderInstance` buffer appended after the scene's own instances, or (b) a `posDrawHook(encoder, uniforms)` that runs inside the main pass with depth bound — the "two-pass rendering" style extension. (b) is more general (custom pipelines, blending); (a) is probably 90% of uses.
3. **Live per-body color/emissive.** `Collider.renderColor` is authored at scene build; interactive apps re-color bodies per frame (button flash, reach highlight pulse, grip state). A small app-updatable color-override buffer sampled by the instance build (index → rgba, alpha 0 = no override), or a third `colorMode` reading such a buffer, would cover it. Emissive currently has no channel at all — `RenderInstance.params` looks like it has no free lane, but `color.w` carries shape type in ≤2 bits of information; a packed emissive there, or a fourth vector, would do.
4. **Visual-mesh authoring question.** `rigidMeshRenderSurface` looks like the intended path for visual-only geometry bound to bodies (we'd use it for a slim arm skeleton instead of showing the fat collision proxies). Is app-side authoring of that surface part of the intended contract, or is it importer-only today? A doc note (or a tiny builder API) would settle it.

None of these are objections to merging — the boundary (`GPUSimRenderableScene`, weak `source`, revision-gated camera adoption) held up well under a real, messy consumer, and the queue-synchronization contract (solver syncs before render, renderer retires before returning) worked at 3 world-ticks-per-frame without artifacts. Plan on our side: once 1–3 land we move FrankaTeleop onto `GPUSimRenderer` outright, and the teleopkit default demo gets rebuilt on `GPUSim` + `GPUSimRenderer` as the showcase.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by LuckyIYI** [comment]
Addendum: `--mechtest` also passed on the consumer port (all mechanism rigs, including the R5 breakaway-bond rig — which exercises the requested `repairJoints()` API directly: its reset re-arm check cannot pass without it). That completes the gate battery on this branch: selftest, remotetest (33 checks), mechtest all green through public products.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by LuckyIYI** [comment]
One more data point: the arm looked capsule-mis-sized in our consumer's renderer view, so we built a measured probe (a static capsule of length 0.4 / radius 0.05 flanked by reference boxes of `length+2r` and bare `length`, side-on orthographic-ish camera, frame captured via `frameCompletionHandler`). Verdict: **the renderer's capsule convention is exactly right** — silhouette equals the `length+2r` box, width equals a 2r sphere — and it matches both the collision kernel (`shape.x * 0.5` segment + caps) and SimCore's documented `size.x = cylinder length`.

The visual difference against our app's built-in view is entirely because GPUSimRenderer draws the true collision proxies (authored fat on purpose) while the app draws a slim visual-only skeleton. That makes request #4 (an app-authoring path for `rigidMeshRenderSurface`-style visual geometry bound to bodies) the one that matters most for making package-rendered scenes presentable, ahead of the color/emissive items.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by LuckyIYI** [comment]
Addressed the consumer-port feedback in `b7b968c`.

- Added public `GPUSolver.brokenJointIndices()` and `repairJoints(_:)`. Inspection excludes disabled non-breakable interaction slots. Repair validates the full explicit index set before mutation, re-arms only fractured breakable joints, clears lambda/motor/dynamics history, and restores authored penalties. Selective/all repair and disabled drag-slot behavior are covered by a Metal-backed test.
- Added app-owned cameras through look-at poses and world-to-view matrices. Camera motion preserves temporal history by default; callers can reset it for cuts, and `useOrbitCamera()` cleanly returns control to the orbit camera.
- Added sparse, live per-body albedo/emissive overrides. They apply to analytic rigid instances, rigid visual meshes, and soft surfaces; invalid indices are safely ignored across scene swaps. Upload work is proportional to active overrides rather than total body count.
- Added a depth-tested, alpha-blended auxiliary primitive pass for boxes, spheres, tori, and capsules, with emission and no shadow/GTAO participation. I chose the bounded instance API over a raw encoder hook so renderer attachment, pipeline, uniform, and synchronization invariants remain internal.
- Confirmed rigid visual meshes are an intended public authoring API and documented the complete replacement path: body-local `SurfaceMesh` + `SceneRigidMesh`, with collision colliders authored using `isRendered: false`. The standalone external consumer now compiles and runs this exact hidden-proxy/visible-mesh arrangement.

Validation:

- Root package: 272 tests passed, 0 failures.
- Development package: 427 tests passed, 20 expected optional MLX/GEAR skips, 0 failures.
- Production SwiftPM build passed.
- Architecture suite: 22 tests passed and dependency-boundary verifier passed.
- Standalone `GPUSimRenderer` consumer built and ran through public products only.
- Metal render/appearance ABI tests passed, and a 30-frame visual smoke run completed cleanly.

@codex review

**Comment by LuckyIYI** [comment]
Re-verified after b7b968c — every review item confirmed working from the external consumer:

- **`brokenJointIndices()` / `repairJoints()`**: consumer builds with no local patches; `--selftest` and `--remotetest` (33 checks) pass on the branch through public products only.
- **`setCamera`**: the game's wrist camera (eye at the lens pod, palm-forward up vector, updated per frame from the palm body pose) works exactly as hoped — textbook egocentric gripper view.
- **`bodyAppearances`**: live per-body albedo+emissive verified (Franka-white styling, dark gripper, a pulsing reach-highlight driven per frame).
- **`auxiliaryInstances`**: we went further than overlays — the arm's contact proxies are now authored `isRendered: false` (the instance builder honors it) and the game's slim visual skeleton (9 capsule bones + joint balls, world-posed per frame) renders through the aux pass. The package-rendered arm is now visually indistinguishable from our bespoke renderer's, which also closes the "capsules look mis-sized" thread: those were the honest contact proxies all along.

Two small follow-ups, neither blocking: an option for auxiliary instances to cast directional shadows (the skeleton arm currently leaves only the gripper's shadow), and a doc note that `GPUSimRenderAppearance` cannot hide a body — `isRendered` at authoring time is the intended path, which is worth stating since it's the first thing a migrating app will look for.

With this, our plan stands: FrankaTeleop's bespoke renderer is now fully replaceable, and the teleopkit demo will be rebuilt on GPUSim + GPUSimRenderer.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by LuckyIYI** [comment]
Found one real aux-pass limitation while using it for the skeleton: `auxiliaryDepthState` has `isDepthWriteEnabled = false`, so auxiliary instances depth-test against the scene but never against **each other** — overlapping opaque-intent instances composite in submission order, and a multi-bone arm literally reads inside-out from most angles. We work around it by sorting instances back-to-front from the camera every frame, which fixes ordering but cannot fix interpenetrating primitives (bone joints).

Suggestion: run `opacity == 1` auxiliary instances through a depth-writing variant (they're opaque by declaration), keeping the blended no-write path for translucent ones. That would make the aux pass a complete solution for solid app geometry, not just wispy overlays — and consumers wouldn't need the CPU sort.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Comment by LuckyIYI** [comment]
Verified the aux depth fix locally: flipping `auxiliaryDepth.isDepthWriteEnabled` to `true` makes the multi-bone skeleton solid from every angle (bones occlude each other correctly; the inside-out effect is gone, no CPU sorting needed). That blanket flip does mean translucent aux instances now also write depth — fine for our use, but the clean version is the opacity partition from the previous comment: depth-writing state for `opacity == 1` instances, the current no-write blended state for the rest. Either way, consider it confirmed that the current always-off depth write makes opaque aux geometry unusable beyond a single instance.

(Also for anyone consuming the package: orbit gestures are app-side — the dev app keeps them, so a bare `configure(view)` window has no camera control until you wire drag/scroll onto the MTKView. A tiny optional `GPUSimOrbitGestureView` in the package, or a doc snippet, would save every consumer the same 40 lines.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: Sources/GPUSimRenderer/GPUSimRenderer.swift [hunk 1]
```diff
@@ -1068,10 +1530,72 @@ final class Renderer: NSObject, MTKViewDelegate {
...
         guard let debugDepthState = device.makeDepthStencilState(
                 descriptor: debugDepth) else {
-            throw RendererInitializationError.depthState("convex-debug")
+            throw GPUSimRendererError.depthState("convex-debug")
         }
         self.debugDepthState = debugDepthState
 
+        let auxiliaryDepth = MTLDepthStencilDescriptor()
+        auxiliaryDepth.depthCompareFunction = .lessEqual
+        auxiliaryDepth.isDepthWriteEnabled = false
```
[the remark this answers is not in the document]
Comment (LuckyIYI) [reply Sources/GPUSimRenderer/GPUSimRenderer.swift]: Fixed in d8f0320. Auxiliary instances are now partitioned by opacity: opaque instances use unblended PBR pipelines with depth writes enabled, while translucent instances keep no-write blending and are sorted back-to-front globally across primitive types. Zero-opacity instances are skipped. Added regression coverage plus the related visibility and orbit-control documentation notes.

**Comment by LuckyIYI** [comment]
Follow-up pushed in `d8f0320`:

- Fully opaque auxiliary instances now render first with depth writes enabled and unblended PBR pipelines, so multi-part solid geometry self-occludes correctly.
- Translucent instances retain the no-depth-write blended path and are sorted back-to-front from the active camera. They are submitted individually so the order remains correct across mixed primitive types.
- Non-positive/invalid opacity is skipped, and the partition/order behavior has regression coverage.
- The docs now state explicitly that `GPUSimRenderAppearance` cannot hide geometry and that `isRendered: false` is the collision-proxy authoring mechanism.
- I documented host-side orbit input mapping instead of adding an AppKit/UIKit policy view to the renderer package.

I kept auxiliary shadow casting out of this correction: it remains explicitly unsupported rather than changing overlay semantics implicitly. If added later, it should be a deliberate per-instance option, not a blanket behavior for all opaque guides.

Validation: 273 root tests passed, the external renderer consumer built and ran, and both production and Development/AVBDApp builds passed.

@codex review

**Comment by LuckyIYI** [comment]
d8f0320 verified from the consumer: the opaque/translucent depth-state partition renders the multi-bone skeleton solid with no CPU sorting, while the translucent target marker and drop beam still blend correctly. Aux pass is now a complete solution for app geometry. 

🤖 Generated with [Claude Code](https://claude.com/claude-code)

File: Makefile [hunk 2]
```diff
@@ -81,20 +112,21 @@ verify-mlx-rl:
...
 verify-release: verify-core verify-mlx-rl app-ml
 
-cli: build
-	@echo "binary: .build/release/avbd"
+cli: workspace-build
+	@echo "binary: $(DEV_BUILD_DIR)/release/avbd"
 
 # Wrap the release executable + owned resource bundles into a double-clickable .app
-app: build
+app: workspace-build
```
[the remark this answers is not in the document]
Comment (LuckyIYI) [reply Makefile:119]: Fixed in fe32f03. The app recipe now creates root .build before mktemp, matching app-ml. Verified against the exact reported state by moving the existing root build cache aside: make app recreated .build, completed the Development release build, signed the bundle, and atomically published AVBD.app. The original cache was then restored unchanged.

**Comment by LuckyIYI** [comment]
P1 clean-checkout packaging fix is in fe32f03 and verified with root .build absent. The app recipe now creates its staging parent before mktemp; make app completed, signed, and atomically published the bundle in that state. @codex review

**Comment by LuckyIYI** [comment]
**Final review — ready to merge.** (GitHub won't let this account formally approve its own PR, so the verdict lands as a comment.) Verified on the current head (2ee2da6, with #12 and #13 integrated and main merged in), all on Apple silicon:

- Root package: `swift build` clean; **273 XCTests + 2 Swift Testing tests, 0 failures**.
- Both integration fixtures build and run (`PackageConsumer`, `RendererPackageConsumer`).
- Development package builds against the split.
- Real-world consumer (the FrankaTeleop game + PuzzleGen + teleopkit as an external SwiftPM package): builds with an import swap only; `--selftest` and `--remotetest` (33 teleop/control checks over live UDP) pass; the packaged renderer drives the live game at ~2.5 ms GPU with phone teleoperation in the loop.
- Renderer output visually verified on the bin scene: content-fitted crisp shadows, auxiliary skeleton casting, AO stable across enable/disable toggles.

API surface after the review rounds: single-import `GPUSim` facade; backend-neutral `GPUSimRenderableScene` with compatibility-defaulted bounds and retirement declarations; app-owned cameras; live per-body appearances; opt-in auxiliary shadow casters; per-joint fracture APIs; no debug scaffolding in the public surface; documentation matches behavior.

🤖 Generated with [Claude Code](https://claude.com/claude-code)
